---
title: "eQTL exploration"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
library(tidyverse)
library(magrittr)
library(parallel)
library(here)
library(SeqArray)
library(AnnotationHub)
library(purrr)
library(ggpubr)
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores <- detectCores() - 2
```

## Annotation

```{r}
ah <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(rdataclass == "EnsDb")
ensDb <- ah[["AH64906"]]
genes <- genes(ensDb)
mcols(genes) <- mcols(genes)[
  c("gene_id", "gene_name", "gene_biotype", "entrezid")
]
exons <- exonsBy(ensDb, by = "gene")
```

## topTable

```{r}
topTable <- read_csv(here("ext_data/FMR1vWildtype.csv"))
```

# GDS file

```{r}
vcfPath <- "/hpcfs/users/a1647910/210305_fmr1_snv/12_selectVariants/mergedVcf/mergedVcf.vcf.gz"
gdsPath <- "/hpcfs/users/a1647910/210305_fmr1_snv/12_selectVariants/gds/mergedGds.gds"
makeGds <-!file.exists(gdsPath)
if (makeGds) {
  dir.create(dirname(gdsPath))
  seqVCF2GDS(vcfPath, gdsPath)
}
```

```{r}
gds <- seqOpen(gdsPath, readonly = FALSE)
```

## Metadata

```{r}
samples <- seqGetData(gds, "sample.id")
## To filter samples by genotype
wt <- str_detect(samples, "A|D|G|L")
mut <- str_detect(samples, "S")
```

## SNPs

```{r}
drChrs <- paste(c(1:25))
seqResetFilter(gds)
snps <- tibble(
  variant.id = seqGetData(gds, "variant.id"),
  chromosome = seqGetData(gds, "chromosome"),
  position = seqGetData(gds, "position"),
  allele = seqGetData(gds, "allele")
) %>%
  dplyr::filter(chromosome %in% drChrs) %>%
  as_tibble()
nucCols <- tibble(A = NA, C = NA, G = NA, T = NA)
```

# TSV files

```{r}
files <- list.files("/hpcfs/users/a1647910/210305_fmr1_snv/13_countAlleles/counts/", full.names = TRUE)
rawTsvs <- lapply(files, read_tsv)
```

# Allele depths

## Genes of interest

```{r}
deGoi <- c("fmr1", "smtnl1", "CU468164.1", "slc7a2", "si:dkey-35h6.1",
           "plekhf1", "ctsf", "rnf4", "cyp4v7", "si:dkey-31g6.6",
           "pcdh1g1", "gla", "si:dkey-85k7.7", "rtkn2a", "slco1d1",
           "znf1179", "cidec", "mgst2", "nocta", "zgc:66433", "atat1")
deGR <- subset(genes, gene_name %in% deGoi)
deTib <- tibble(
  symbol = mcols(deGR)$gene_name,
  chromosome = as.vector(seqnames(deGR)),
  start = start(deGR),
  end = end(deGR),
  geneId = mcols(deGR)$gene_id
) %>%
  dplyr::filter(chromosome %in% drChrs) %>% 
  dplyr::arrange(chromosome, start)
```

```{r}
deChr14 <- dplyr::filter(deTib, chromosome == 14) %>%
  pull(symbol)
deOther <- dplyr::filter(deTib, chromosome != 14) %>%
  pull(symbol)
```

```{r}
## Sample 10 non-DE genes in chr 14 with high p-value and expression,
## that are positioned closely to the DE genes.
genesByLoc <- topTable %>% 
  dplyr::select(
    gene_id = ensembl_gene_id,
    gene_name = external_gene_name,
    logFC, AveExpr, P.Value, adj.P.Val
  ) %>% 
  right_join(as_tibble(genes)) %>%
  dplyr::arrange(seqnames, start) %>%
  dplyr::filter(seqnames %in% drChrs)
set.seed(17)
ndeChr14 <- lapply(deChr14, function(x) {
  row <- which(genesByLoc$gene_name == x)
  rowLeft <- row - 25
  rowRight <- row + 25
  dplyr::slice(genesByLoc, rowLeft:rowRight) %>%
    dplyr::filter(
      seqnames == 14, 
      adj.P.Val > 0.80,
      AveExpr > 2
    ) %>%
    pull(gene_name) %>%
    sample(1)
})
```

## Get counts from GDS

```{r}
minCounts <- 0
geneName <- "smtnl1"

sym <- geneName
geneId <- genes %>%
  subset(seqnames %in% drChrs) %>%
  subset(gene_name == sym) %>%
  .$gene_id
eoi <- exons[[geneId]]

## Get variants IDs that lie only in exons
seqSetFilter(gds, eoi)
vIds <- seqGetData(gds, "variant.id")
seqResetFilter(gds)

gdsCounts <-lapply(samples, function(sample){
  sampleSel <- samples == sample
  lapply(vIds, function(vId){
    seqResetFilter(gds)
    seqSetFilter(gds, variant.sel = vId, sample.sel = sampleSel)
    als <- seqGetData(gds, "allele") %>%
      str_split(",") %>% 
      unlist()
    seqGetData(gds, "annotation/format/AD")$data %>%
      t() %>%
      rowSums(na.rm = TRUE) %>%
      set_names(als) %>%
      bind_rows() %>%
      mutate(variant.id = vId)
  }) %>%
    purrr::reduce(full_join) %>%
    add_column(!!!nucCols[!names(nucCols) %in% names(.)]) %>%
    left_join(snps, by = "variant.id") %>%
    dplyr::select(
      variant.id, chromosome, position, allele, 
      A, C, G, T, everything()
    ) %>%
    pivot_longer(
      cols = !c("variant.id", "chromosome", "position", "allele"),
      names_to = "base"
    ) %>%
    na.omit() %>%
    mutate(gene = sym) %>%
    dplyr::select(gene, chromosome, position, alleleGDS = allele, base, valueGDS = value)
})
```

## Get counts from tsv

```{r}
files <- list.files("/hpcfs/users/a1647910/210305_fmr1_snv/13_countAlleles/counts/", full.names = TRUE)
tsvCounts <- lapply(rawTsvs, function(x){
  x %>%
    mutate(
      alleleTSV = paste0(REF_NUCLEOTIDE, ",", ALT_NUCLEOTIDE),
      CONTIG = as.character(CONTIG)
    ) %>%
    pivot_longer(
      cols = c("REF_NUCLEOTIDE", "ALT_NUCLEOTIDE"),
      names_to = "variant",
      values_to = "base"
    ) %>% 
    mutate(value = ifelse(variant == "REF_NUCLEOTIDE", REF_COUNT, ALT_COUNT)) %>%
    dplyr::select(chromosome = CONTIG, position = POSITION, alleleTSV, base, valueTSV = value)
})
```

## Compare counts

```{r}
comp <- lapply(1:8, function(x){
  chrFilt <- gdsCounts[[x]]$chromosome %>%
    unique()
  posFilt <- gdsCounts[[x]]$position
  dplyr::filter(
    tsvCounts[[x]],
    chromosome == chrFilt,
    position %in% posFilt,
    base != "N"
  ) %>%
    full_join(gdsCounts[[x]]) %>%
    dplyr::select(chromosome, position, alleleGDS, alleleTSV, base, valueGDS, valueTSV) %>%
    dplyr::arrange(position)
})
```

